### PR TITLE
Router update

### PR DESCRIPTION
**Pull request by CJPoll**

### CODE REVIEW

File: lib/mix/tasks/schema.create.ex [hunk 1]
```diff
@@ -24,7 +24,7 @@ defmodule Mix.Tasks.Schema.Create do
   end
 
   def create_schema(subject, schema) do
-    Logger.info("Creating schema for subjejct: #{subject}")
+    Logger.info("Creating schema for subject: #{subject}")
```
Comment (CJPoll) [inline lib/mix/tasks/schema.create.ex]: Spelling fix

File: lib/rivulet/application.ex [hunk 2]
```diff
@@ -26,11 +27,16 @@ defmodule Rivulet.Application do
     kafka_hosts = kafka_brokers()
 
     if System.get_env("MIX_ENV") != "test" do
-      client_name = Application.get_env(:rivulet, :publish_client_name)
-      unless client_name do
-        raise "Application.get_env(:rivulet, :publish_client_name) not configured"
-      end
-      :ok = :brod.start_client(kafka_hosts, client_name, _client_config=[auto_start_producers: true])
+      client_name = Rivulet.client_name
```
Comment (CJPoll) [inline lib/rivulet/application.ex]: Add some functions for easily getting the client (for both publishing and consuming)

File: lib/rivulet/application.ex [hunk 3]
```diff
@@ -26,11 +27,16 @@ defmodule Rivulet.Application do
...
+
+      default_producer_config = [max_batch_size: 100]
```
Comment (CJPoll) [inline lib/rivulet/application.ex]: Set a default max batch size to prevent "MessageTooLarge" exceptions

File: lib/rivulet/kafka/consumer.ex [hunk 4]
```diff
@@ -11,8 +11,7 @@ defmodule Rivulet.Consumer do
         offset_commit_policy: :commit_to_kafka_v2,
         offset_commit_interval_seconds: 5
       ],
-      consumer_config: [begin_offset: :earliest],
-      message_type: :message_set
```
Comment (CJPoll) [inline lib/rivulet/kafka/consumer.ex]: We never use anything but `:message_set`, and I don't want our engineers to use `:message`. If they really only care about 1 at a time, they can just Enum.map over the message set.

File: lib/rivulet/kafka/publisher.ex [hunk 5]
```diff
@@ -61,64 +60,24 @@ defmodule Rivulet.Kafka.Publisher do
...
-        (:error) -> :error
-        ({:still_running, task}) -> task
+        ({:ok, _call_ref}) -> :ok
+        ({:error, _reason} = err) ->
+          Logger.error("Bulk publishing failed: #{inspect err}")
+          :error
       end)
-      |> do_wait(counter - 1)
-    end
-  end
```
Comment (CJPoll) [inline lib/rivulet/kafka/publisher.ex]: We don't need to manage parallelism when publishing ourselvs. Just don't.

File: lib/rivulet/kafka/router.ex [hunk 6]
```diff
@@ -36,126 +37,59 @@ defmodule Rivulet.Router do
...
 
+    {consumer_group, stream_topics, streams}
+  end
+
+  defmacro __before_compile__(env) do
+    {consumer_group, stream_topics, streams} = get_stuff(env)
+
     quote do
       def start_link do
-        require Logger
```
Comment (CJPoll) [inline lib/rivulet/kafka/router.ex]: Extract some of this stuff out to functions for better stacktraces.

File: lib/rivulet/kafka/router.ex [hunk 7]
```diff
@@ -36,126 +37,59 @@ defmodule Rivulet.Router do
...
-
-        Rivulet.Consumer.start_link(__MODULE__, config)
+        Rivulet.Kafka.Router.Funcs.start_link(__MODULE__, unquote(consumer_group), unquote(stream_topics))
       end
 
       def init(_) do
         {:ok, {}}
       end
 
-      def handle_messages(%Rivulet.Kafka.Partition{topic: topic} = partition, messages, state) do
```
Comment (CJPoll) [inline lib/rivulet/kafka/router.ex]: More extractions for better stacktraces

File: lib/rivulet/kafka/router.ex [hunk 8]
```diff
@@ -36,126 +37,59 @@ defmodule Rivulet.Router do
...
+      def handle_messages(partition, messages, state) do
+        Rivulet.Kafka.Router.Funcs.handle_messages(partition, messages, unquote(streams))
 
         {:ok, :ack, state}
       end
     end
   end
 
-  defmacro defsource(topic, [do: routes]) do
+  defmacro defstream(topic, [do: routes]) do
```
Comment (CJPoll) [inline lib/rivulet/kafka/router.ex]: Rename from `defsource` to `defstream`

File: lib/rivulet/kafka/router/funcs.ex [hunk 9]
```diff
@@ -0,0 +1,105 @@
+defmodule Rivulet.Kafka.Router.Funcs do
```
Comment (CJPoll) [inline lib/rivulet/kafka/router/funcs.ex]: We extracted generated code out to here for better stacktraces. Generally the same as before, but a few minor differences (like removing the option for `:message`)

File: lib/rivulet/kafka/transformer/avro_inspect.ex [hunk 10]
```diff
@@ -0,0 +1,11 @@
+defmodule Rivulet.Transformer.AvroInspect do
```
Comment (CJPoll) [inline lib/rivulet/kafka/transformer/avro_inspect.ex]: For debugging in Dev

File: lib/rivulet/kafka/transformer/inspect.ex [hunk 11]
```diff
@@ -0,0 +1,6 @@
+defmodule Rivulet.Transformer.Inspect do
```
Comment (CJPoll) [inline lib/rivulet/kafka/transformer/inspect.ex]: For debugging in Dev

File: lib/rivulet/kafka/transformer/pass_through.ex [hunk 12]
```diff
@@ -0,0 +1,5 @@
+defmodule Rivulet.Transformer.PassThrough do
```
Comment (CJPoll) [inline lib/rivulet/kafka/transformer/pass_through.ex]: Useful for repartitioning

File: mix.exs [hunk 13]
```diff
@@ -35,7 +35,7 @@ defmodule Rivulet.Mixfile do
 
   defp deps do
     [{:dialyxir, "~> 0.5.0", only: [:dev, :test], runtime: false},
-     {:avro_ex, "~> 0.1.0-beta.1"},
+     {:avro_ex, "~> 0.1.0-beta.2"},
```
Comment (CJPoll) [inline mix.exs]: Use latest avro_ex for better error messages.

File: Dockerfile [hunk 14]
```diff
@@ -1,4 +1,4 @@
-FROM elixir:1.4
+FROM elixir:1.6
```
Comment (aweagel) [inline Dockerfile]: 👏

File: docker-compose.yml [hunk 15]
```diff
@@ -1,7 +1,55 @@
...
+    command: echo "Not an app"
+    environment:
+      HOSTNAME: host1
+    depends_on:
+      - kafka
+      - schema-registry
     volumes:
       - .:/app
+  kafka:
+    image: confluentinc/cp-kafka:3.3.0
```
Comment (aweagel) [inline docker-compose.yml]: This should be `4.0.0` in order to run off kafka 1.0+ by default

File: docker-compose.yml [hunk 16]
```diff
@@ -1,7 +1,55 @@
...
+    environment:
+      KAFKA_ZOOKEEPER_CONNECT: "zookeeper:2181"
+      KAFKA_ADVERTISED_LISTENERS: PLAINTEXT://kafka:9092
+      KAFKA_OFFSETS_TOPIC_REPLICATION_FACTOR: 1
+      KAFKA_DELETE_TOPIC_ENABLE: "true"
+    volumes:
+      - kafka:/var/lib/kafka/data
+    restart: on-failure
+  zookeeper:
+    image: confluentinc/cp-zookeeper:3.3.0
```
Comment (aweagel) [inline docker-compose.yml]: This should be `4.0.0` in order to run off kafka 1.0+ by default

File: docker-compose.yml [hunk 17]
```diff
@@ -1,7 +1,55 @@
...
+    volumes:
+      - zookeeper:/var/lib/zookeeper/data
+    environment:
+      ZOOKEEPER_CLIENT_PORT: 2181
+  schema-registry:
+    image: confluentinc/cp-schema-registry:3.3.0
```
Comment (aweagel) [inline docker-compose.yml]: This should be `4.0.0` in order to run off kafka 1.0+ by default

File: lib/rivulet/application.ex (hunk 2 above)
Comment (aweagel) [reply lib/rivulet/application.ex]: I like this idea 👍

File: lib/rivulet/application.ex [hunk 18]
```diff
@@ -26,11 +27,16 @@ defmodule Rivulet.Application do
...
+
+      :ok =
+        :brod.start_client(kafka_hosts, client_name, _client_config=[
+          auto_start_producers: true,
+          allow_topic_auto_creation: false,
```
Comment (aweagel) [inline lib/rivulet/application.ex]: Thank you! 🙏

File: lib/rivulet/kafka/consumer.ex (hunk 4 above)
Comment (aweagel) [reply lib/rivulet/kafka/consumer.ex]: I agree 👍

File: lib/rivulet/kafka/publisher.ex (hunk 5 above)
Comment (aweagel) [reply lib/rivulet/kafka/publisher.ex]: 🙌

File: config/config.exs [hunk 19]
```diff
@@ -1,9 +1,6 @@
 use Mix.Config
 
 config :rivulet,
-  avro_schema_registry_uri: %URI{scheme: "http", host: "localhost", port: 8081},
-  kafka_brokers: [localhost: 9092],
-  consumer_group: "rivulet",
-  kafka_version: "0.11.0",
-  use_ssl: false,
-  publish_client_name: :"rivulet_brod_client-#{System.get_env("HOSTNAME")}"
```
Comment (aweagel) [inline config/config.exs]: I'd like to better understand the reasoning for the changes here

File: lib/rivulet/kafka/router.ex (hunk 8 above)
Comment (aweagel) [reply lib/rivulet/kafka/router.ex]: 🌊

**Pull request merged**